server/src/models/timestamps.rs

```diff
@@ -0,0 +1,352 @@
+
+        match delta_of_delta {
+            0 => compressed_timestamps.append_a_zero_bit(),
+            -63..=64 => {
+                compressed_timestamps.append_bits(0b10, 2);
+                compressed_timestamps.append_bits(delta_of_delta as u32, 7);
+            }
+            -255..=256 => {
+                compressed_timestamps.append_bits(0b110, 3);
+                compressed_timestamps.append_bits(delta_of_delta as u32, 9);
+            }
+            -2047..=2048 => {
+                compressed_timestamps.append_bits(0b1110, 4);
+                compressed_timestamps.append_bits(delta_of_delta as u32, 12);
+            }
+            _ => {
+                compressed_timestamps.append_bits(0b1111, 4);
+                compressed_timestamps.append_bits(delta_of_delta as u32, 32);
+            }
+        }
+        last_delta = delta;
+        last_timestamp = *timestamp;
+    }
+
+    // All remaining bits in the byte the BitVecBuilder is currently packing
+    // bits into is set to one to indicate that all timestamps are decompressed.
+    compressed_timestamps.finish_with_one_bits()
+}
+
+/// Decompress all of a segment's timestamps which are compressed as
+/// `start_time` for segments of length one, `start_time` and `end_time` for
+/// segments of length two, the segment's length for regular time series, or
+/// using Gorilla's compression method for timestamps for irregular time series.
+pub fn decompress_all_timestamps(
+    start_time: Timestamp,
+    end_time: Timestamp,
+    residual_timestamps: &[u8],
+) -> TimestampArray {
+    if residual_timestamps.is_empty() && start_time == end_time {
+        // Timestamps are assumed to be unique so the segment has one timestamp.
+        let mut timestamp_builder = TimestampBuilder::new(1);
+        timestamp_builder.append_value(start_time);
+        timestamp_builder.finish()
+    } else if residual_timestamps.is_empty() {
+        // Timestamps are assumed to be unique so the segment has two timestamp.
+        let mut timestamp_builder = TimestampBuilder::new(2);
+        timestamp_builder.append_value(start_time);
+        timestamp_builder.append_value(end_time);
+        timestamp_builder.finish()
+    } else if residual_timestamps[0] & 128 == 0 {
+        // The flag bit is zero, so only the segment's length is stored as an
+        // integer with all the prefix zeros stripped from the integer.
+        decompress_all_regular_timestamps(start_time, end_time, residual_timestamps)
+    } else {
+        // The flag bit is one, so the timestamps are compressed as
+        // delta-of-deltas stored using a variable length binary encoding.
+        decompress_all_irregular_timestamps(start_time, end_time, residual_timestamps)
+    }
+}
+
+/// Decompress all of a segment's timestamps, which for this segment are sampled
+/// at a regular sampling interval, and thus compressed as the segment's length.
+fn decompress_all_regular_timestamps(
```

Comment:
Maybe add some whitespace to this function.

Reply:
Fixed.